src/components/tl/ucp/allgather/allgather_bruck.c

```diff
@@ -0,0 +1,254 @@
+    ucc_status_t       global_status = UCC_OK;
+    ucc_status_t       status;
+
+    tl_trace(UCC_TASK_LIB(task), "ucc_tl_ucp_allgather_bruck_finalize");
+
+    if (task->allgather_bruck.scratch_header != NULL) {
+        /* deallocate scratch buffer */
+        global_status = ucc_mc_free(task->allgather_bruck.scratch_header);
+        if (ucc_unlikely(global_status != UCC_OK)) {
+            tl_error(UCC_TASK_LIB(task),
+                     "failed to free scratch buffer memory");
+        }
+        task->allgather_bruck.scratch_size = 0;
+    }
+
+    status = ucc_tl_ucp_coll_finalize(&task->super);
+    if (ucc_unlikely(status != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task),
+                 "failed to finalize allgather bruck collective");
+        global_status = status;
+    }
+    return global_status;
+}
+
+/* Inspired by implementation: https://github.com/open-mpi/ompi/blob/main/ompi/mca/coll/base/coll_base_allgather.c */
+void ucc_tl_ucp_allgather_bruck_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t      *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t      *team = TASK_TEAM(task);
+    ucc_rank_t              trank = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t              tsize = UCC_TL_TEAM_SIZE(team);
+    void                   *rbuf  = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t       rmem  = TASK_ARGS(task).dst.info.mem_type;
+    ucc_datatype_t          dt    = TASK_ARGS(task).dst.info.datatype;
+    size_t                  count = TASK_ARGS(task).dst.info.count;
+    ucc_mc_buffer_header_t *scratch_header =
+        task->allgather_bruck.scratch_header;
+    size_t       scratch_size = task->allgather_bruck.scratch_size;
+    size_t       data_size    = (count / tsize) * ucc_dt_size(dt);
+    ucc_rank_t   recvfrom, sendto;
+    ucc_status_t status;
+    size_t       blockcount, distance;
+    void        *tmprecv, *tmpsend;
+
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return;
+    }
+
+    /* On each step doubles distance */
+    distance = 1 << task->tagged.send_posted;
+    tmpsend  = rbuf;
+    while (distance < tsize) {
+
+        recvfrom = (trank + distance) % tsize;
+        sendto   = (trank + tsize - distance) % tsize;
+
+        tmprecv = PTR_OFFSET(tmpsend, distance * data_size);
+
+        if (distance <= tsize >> 1) {
+            blockcount = distance;
+        } else {
+            /* send-recv all reminder*/
+            blockcount = tsize - distance;
+        }
+
+        /* Sendreceive */
+        UCPCHECK_GOTO(ucc_tl_ucp_send_nb(tmpsend, blockcount * data_size, rmem,
+                                         sendto, team, task),
+                      task, out);
+        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(tmprecv, blockcount * data_size, rmem,
+                                         recvfrom, team, task),
+                      task, out);
+
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
```

Comment:
why need to wait for both send and recv to complete? I think you can start next send once prev recv is done

Reply:
I've tried to do it, but wasn't able to notice any gains on my setup.

```
mpirun -x UCC_TL_UCP_TUNE=allgather:0-inf:@3 --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=/home/ikryukov/work/ucc/install/lib:/home/ikryukov/work/ucx/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCC_CL_BASIC_TLS=ucp -np 16 ./install/bin/ucc_perftest -c allgather -F -b 1 -e 4k
Collective:             Allgather
Memory type:            host
Datatype:               float32
Reduction:              N/A
Inplace:                0
Warmup:
  small                 100
  large                 20
Iterations:
  small                 1000
  large                 200

       Count        Size                Time, us                           Bandwidth, GB/s
                                 avg         min         max         avg         max         min
           1           4        1.60        1.56        1.65        0.04        0.04        0.04
           2           8        1.83        1.61        1.91        0.07        0.07        0.06
           4          16        2.04        1.80        2.36        0.12        0.13        0.10
           8          32        2.46        2.38        2.55        0.20        0.20        0.19
          16          64        2.93        2.87        2.98        0.33        0.33        0.32
          32         128        3.51        3.44        3.57        0.55        0.56        0.54
          64         256        3.77        3.63        3.89        1.02        1.06        0.99
         128         512        5.47        5.18        5.72        1.40        1.48        1.34
         256        1024       12.50       10.35       13.38        1.23        1.48        1.15
         512        2048       20.50       16.25       23.01        1.50        1.89        1.34
        1024        4096       34.41       22.96       39.38        1.79        2.68        1.56
        2048        8192       60.71       35.99       70.53        2.02        3.41        1.74
        4096       16384      113.12       58.69      131.59        2.17        4.19        1.87

==================
After
==================
mpirun -x UCC_TL_UCP_TUNE=allgather:0-inf:@3 --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=/home/ikryukov/work/ucc/install/lib:/home/ikryukov/work/ucx/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCC_CL_BASIC_TLS=ucp -np 16 ./install/bin/ucc_perftest -c allgather -F -b 1 -e 4k
Collective:             Allgather
Memory type:            host
Datatype:               float32
Reduction:              N/A
Inplace:                0
Warmup:
  small                 100
  large                 20
Iterations:
  small                 1000
  large                 200

       Count        Size                Time, us                           Bandwidth, GB/s
                                 avg         min         max         avg         max         min
           1           4        1.60        1.53        1.66        0.04        0.04        0.04
           2           8        1.56        1.51        1.61        0.08        0.08        0.07
           4          16        1.86        1.72        1.97        0.13        0.14        0.12
           8          32        2.45        2.40        2.52        0.20        0.20        0.19
          16          64        2.91        2.82        2.99        0.33        0.34        0.32
          32         128        3.53        3.46        3.61        0.54        0.55        0.53
          64         256        3.75        3.66        3.91        1.02        1.05        0.98
         128         512        5.72        5.39        6.04        1.34        1.43        1.27
         256        1024       12.60       10.37       13.57        1.22        1.48        1.13
         512        2048       20.88       15.93       22.63        1.47        1.93        1.36
        1024        4096       32.32       26.55       37.26        1.90        2.31        1.65
        2048        8192       54.53       39.76       66.31        2.25        3.09        1.85
        4096       16384      100.92       77.76      124.66        2.44        3.16        1.97`

Reply:
Thanks, I ran it on bigger scale and also don't see big difference compared to prev version